lib/core/src/Cardano/Wallet/Address/Pool.hs

```diff
@@ -0,0 +1,216 @@
+    ( AddressState (..) )
+import Control.DeepSeq
+    ( NFData )
+import Data.Map.Strict
+    ( Map )
+import Data.Ord
+    ( Down (..) )
+import Fmt
+    ( Buildable (..) )
+import GHC.Generics
+    ( Generic )
+
+{- HLINT ignore "Avoid restricted qualification" -}
+import qualified Data.List as List
+import qualified Data.Map.Strict as Map
+
+{-------------------------------------------------------------------------------
+    Address Pool, abstract data type
+-------------------------------------------------------------------------------}
+-- | An address pool caches a collection of addresses (type @addr@)
+-- which are derived from a numeric index (type @ix@).
+data Pool addr ix = Pool
+    { generator :: ix -> addr
+    -- ^ Each address is obtained from a numeric index.
+    -- The purpose of the 'Pool' data structure is to (partially)
+    -- cache this mapping,
+    -- because it is expensive to compute (hashing)
+    -- and its inverse is practically impossible to compute.
+    , gap :: Int
+    -- ^ The pool gap determines how 'Used' and 'Unused'
+    -- have to be distributed.
+    -- See 'prop_gap' and 'prop_fresh'.
+    , addresses :: Map addr (ix, AddressState)
+    -- ^ Partial, cached inverse of the 'generator'.
+    -- This map contains all cached addresses @addr@,
+    -- their corresponding indices @ix@,
+    -- and whether they are 'Used' or 'Unused'.
+    -- See 'prop_sequence'.
+    } deriving (Generic)
+
+instance (NFData addr, NFData ix) => NFData (Pool addr ix)
+
+-- | Internal invariant:
+-- The indices of the addresses in a pool form a finite
+-- sequence beginning with 'fromEnum'@ 0@.
+prop_sequence :: (Ord ix, Enum ix) => Pool addr ix -> Bool
+prop_sequence Pool{addresses} =
+    indices `List.isPrefixOf` [toEnum 0..]
+  where
+    indices = List.sort $ map fst $ Map.elems addresses
+
+-- | Internal invariant:
+-- If we order the 'addresses' by their indices,
+-- then there are always /less than/ 'gap' many 'Unused'
+-- addresses between two consecutive 'Used' addresses.
+prop_gap :: Ord ix => Pool addr ix -> Bool
+prop_gap Pool{gap,addresses}
+    = all ((< gap) . length)
+    . filter isUnused
+    . dropFresh
+    $ List.group statuses
+  where
+    isUnused (Unused:_) = True
+    isUnused _ = False
+    dropFresh = drop 1 -- drop items that are checked by 'prop_fresh'.
+    statuses = map snd $ List.sortOn (Down . fst) $ Map.elems addresses
+
+-- | Internal invariant:
+-- If we order the 'addresses' by their indices,
+-- there are exactly 'gap' many 'Unused' addresses after the last
+-- 'Used' address.
+prop_fresh :: Ord ix => Pool addr ix -> Bool
+prop_fresh Pool{gap,addresses} =
+    takeWhile (== Unused) end == replicate gap Unused
+  where
+    end = map snd $ List.sortOn (Down . fst) $ Map.elems addresses
+
+-- | Internal invariant:
+-- All 'addresses' in the pool have been generated from their index
+-- via the pool 'generator'.
+prop_generator :: Eq addr => Pool addr ix -> Bool
```

Comment:
indeed, there should be correspondence between `generate` put into work and what is inside `addresses`, and nothing extra in addresses :+1: